File: tools/source/sdks/c-sdk/src/game_agent_engine_client.c
#include "game_agent_engine_client.h"

#include <stdio.h>

static char gae_buffer[1024];

const char* gae_health_path(void) { return "/health"; }
const char* gae_version_path(void) { return "/api/v1/version"; }
const char* gae_invoke_path(void) { return "/api/v1/invoke"; }
const char* gae_interpret_player_input_path(void) { return "/api/v1/player/input/interpret"; }
const char* gae_advance_tick_path(const char* world_id) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/worlds/%s/ticks/advance", world_id);
    return gae_buffer;
}
const char* gae_world_settings_path(const char* world_id) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/worlds/%s/settings", world_id);
    return gae_buffer;
}
const char* gae_state_components_path(const char* world_id) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/worlds/%s/state-components", world_id);
    return gae_buffer;
}
const char* gae_state_component_path(const char* world_id, const char* component_type) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/worlds/%s/state-components/%s", world_id, component_type);
    return gae_buffer;
}
const char* gae_timelines_path(const char* world_id, int limit) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/worlds/%s/timelines?limit=%d", world_id, limit);
    return gae_buffer;
}
const char* gae_latest_timeline_path(const char* world_id) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/worlds/%s/timelines/latest", world_id);
    return gae_buffer;
}
const char* gae_logs_path(const char* world_id, int limit, int offset, const char* task_type) {
    snprintf(
        gae_buffer,
        sizeof(gae_buffer),
        "/api/v1/logs?world_id=%s&limit=%d&offset=%d%s%s",
        world_id,
        limit,
        offset,
        (task_type && task_type[0]) ? "&task_type=" : "",
        (task_type && task_type[0]) ? task_type : ""
    );
    return gae_buffer;
}
const char* gae_debug_traces_path(const char* world_id, int limit) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/debug/traces?world_id=%s&limit=%d", world_id, limit);
    return gae_buffer;
}
const char* gae_world_policy_path(const char* world_id) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/worlds/%s/policy", world_id);
    return gae_buffer;
}
const char* gae_pending_tasks_path(const char* consumer, int limit) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/runtime/tasks/pending?consumer=%s&limit=%d", consumer, limit);
    return gae_buffer;
}
const char* gae_runtime_tasks_path(const char* category, const char* status, int limit) {
    snprintf(
        gae_buffer,
        sizeof(gae_buffer),
        "/api/v1/runtime/tasks?limit=%d%s%s%s%s",
        limit,
        (category && category[0]) ? "&category=" : "",
        (category && category[0]) ? category : "",
        (status && status[0]) ? "&status=" : "",
        (status && status[0]) ? status : ""
    );
    return gae_buffer;
}
const char* gae_runtime_task_path(const char* task_id) {
    snprintf(gae_buffer, sizeof(gae_buffer), "/api/v1/runtime/tasks/%s", task_id);
    return gae_buffer;
}
const char* gae_claim_runtime_task_payload(const char* task_id, const char* consumer, const char* owner) {
    snprintf(gae_buffer, sizeof(gae_buffer), "{\"task_id\":\"%s\",\"consumer\":\"%s\",\"lease_owner\":\"%s\"}", task_id, consumer, owner);
    return gae_buffer;
}
const char* gae_start_runtime_task_payload(const char* task_id, const char* lease_token) {
    snprintf(gae_buffer, sizeof(gae_buffer), "{\"task_id\":\"%s\",\"lease_token\":\"%s\"}", task_id, lease_token);
    return gae_buffer;
}
const char* gae_heartbeat_runtime_task_payload(const char* task_id, const char* lease_token) {
    snprintf(gae_buffer, sizeof(gae_buffer), "{\"task_id\":\"%s\",\"lease_token\":\"%s\"}", task_id, lease_token);
    return gae_buffer;
}
const char* gae_release_runtime_task_payload(const char* task_id, const char* lease_token, const char* error_message) {
    snprintf(gae_buffer, sizeof(gae_buffer), "{\"task_id\":\"%s\",\"lease_token\":\"%s\",\"error_message\":\"%s\"}", task_id, lease_token, error_message ? error_message : "");
    return gae_buffer;
}
const char* gae_requeue_runtime_task_payload(const char* task_id, int retry_delay_ms, const char* error_message) {
    snprintf(gae_buffer, sizeof(gae_buffer), "{\"task_id\":\"%s\",\"retry_delay_ms\":%d,\"error_message\":\"%s\"}", task_id, retry_delay_ms, error_message ? error_message : "");
    return gae_buffer;
}
const char* gae_runtime_task_stats_path(void) {
    return "/api/v1/runtime/tasks/stats";
}
const char* gae_callback_payload(const char* callback_id, const char* status, const char* result_json) {
    snprintf(gae_buffer, sizeof(gae_buffer), "{\"callback_id\":\"%s\",\"status\":\"%s\",\"result\":%s}", callback_id, status, result_json);
    return gae_buffer;
}

File: tools/source/sdks/java-sdk/src/GameAgentEngineClient.java
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.List;

public class GameAgentEngineClient {
    private final String baseUrl;
    private final String apiKey;
    private final HttpClient httpClient = HttpClient.newHttpClient();

    public GameAgentEngineClient(String baseUrl, String apiKey) {
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
    }

    public String get(String path) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(baseUrl + path))
            .header("X-API-Key", apiKey)
            .header("Accept", "application/json")
            .GET()
            .build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString()).body();
    }

    public String post(String path, String jsonBody) throws Exception {
        return post(path, jsonBody, null);
    }

    public String post(String path, String jsonBody, String callbackRequestId) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(baseUrl + path))
            .header("X-API-Key", apiKey)
            .header("Accept", "application/json")
            .header("Content-Type", "application/json");
        if (callbackRequestId != null && !callbackRequestId.isBlank()) {
            builder.header("X-Callback-Request-Id", callbackRequestId);
        }
        HttpRequest request = builder
            .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
            .build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString()).body();
    }

    public String postOrPut(String method, String path, String jsonBody) throws Exception {
        var builder = HttpRequest.newBuilder()
            .uri(URI.create(baseUrl + path))
            .header("X-API-Key", apiKey)
            .header("Accept", "application/json")
            .header("Content-Type", "application/json");
        HttpRequest request = builder
            .method(method, HttpRequest.BodyPublishers.ofString(jsonBody))
            .build();
        return httpClient.send(request, HttpResponse.BodyHandlers.ofString()).body();
    }

    private String buildQuery(String[][] pairs) {
        List<String> items = new ArrayList<>();
        for (String[] pair : pairs) {
            if (pair == null || pair.length < 2) {
                continue;
            }
            var key = pair[0];
            var value = pair[1];
            if (key == null || value == null || value.isBlank()) {
                continue;
            }
            items.add(URLEncoder.encode(key, StandardCharsets.UTF_8) + "=" + URLEncoder.encode(value, StandardCharsets.UTF_8));
        }
        if (items.isEmpty()) {
            return "";
        }
        return "?" + String.join("&", items);
    }

    public String health() throws Exception { return get("/health"); }
    public String version() throws Exception { return get("/api/v1/version"); }
    public String invoke(String jsonBody) throws Exception { return post("/api/v1/invoke", jsonBody); }
    public String interpretPlayerInput(String jsonBody) throws Exception { return post("/api/v1/player/input/interpret", jsonBody); }
    public String advanceTick(String worldId, String jsonBody) throws Exception {
        return post("/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/ticks/advance", jsonBody);
    }
    public String getWorldSettings(String worldId) throws Exception {
        return get("/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/settings");
    }
    public String setWorldSettings(String worldId, String jsonBody) throws Exception {
        return postOrPut("PUT", "/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/settings", jsonBody);
    }
    public String getStateComponents(String worldId) throws Exception {
        return get("/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/state-components");
    }
    public String getStateComponent(String worldId, String componentType) throws Exception {
        return get("/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/state-components/" + URLEncoder.encode(componentType, StandardCharsets.UTF_8));
    }
    public String putStateComponent(String worldId, String componentType, String jsonBody) throws Exception {
        return postOrPut("PUT", "/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/state-components/" + URLEncoder.encode(componentType, StandardCharsets.UTF_8), jsonBody);
    }
    public String getTimelines(String worldId, int limit) throws Exception {
        return get("/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/timelines" + buildQuery(new String[][]{
            {"limit", limit > 0 ? Integer.toString(limit) : null},
        }));
    }
    public String getLatestTimeline(String worldId) throws Exception {
        return get("/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/timelines/latest");
    }
    public String getLogs(String worldId, int limit, int offset, String taskType) throws Exception {
        return get("/api/v1/logs" + buildQuery(new String[][]{
            {"world_id", worldId},
            {"limit", limit > 0 ? Integer.toString(limit) : null},
            {"offset", offset > 0 ? Integer.toString(offset) : null},
            {"task_type", taskType},
        }));
    }
    public String getDebugTraces(String worldId, int limit) throws Exception {
        return get("/debug/traces" + buildQuery(new String[][]{
            {"world_id", worldId},
            {"limit", limit > 0 ? Integer.toString(limit) : null},
        }));
    }
    public String getWorldPolicy(String worldId) throws Exception {
        return get("/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/policy");
    }
    public String setWorldPolicy(String worldId, String jsonBody) throws Exception {
        return postOrPut("PUT", "/api/v1/worlds/" + URLEncoder.encode(worldId, StandardCharsets.UTF_8) + "/policy", jsonBody);
    }
    public String listPendingRuntimeTasks(String consumer, int limit) throws Exception {
        return get("/api/v1/runtime/tasks/pending?consumer=" + URLEncoder.encode(consumer, StandardCharsets.UTF_8) + "&limit=" + limit);
    }
    public String listRuntimeTasks(String category, String status, int limit) throws Exception {
        return get("/api/v1/runtime/tasks" + buildQuery(new String[][]{
            {"category", category},
            {"status", status},
            {"limit", Integer.toString(limit)},
        }));
    }
    public String getRuntimeTask(String taskId) throws Exception {
        return get("/api/v1/runtime/tasks/" + URLEncoder.encode(taskId, StandardCharsets.UTF_8));
    }
    public String claimRuntimeTask(String taskId, String consumer, String owner) throws Exception {
        return post("/api/v1/runtime/tasks/claim", "{\"task_id\":\"" + taskId + "\",\"consumer\":\"" + consumer + "\",\"lease_owner\":\"" + owner + "\"}");
    }
    public String startRuntimeTask(String taskId, String leaseToken) throws Exception {
        return post("/api/v1/runtime/tasks/start", "{\"task_id\":\"" + taskId + "\",\"lease_token\":\"" + leaseToken + "\"}");
    }
    public String heartbeatRuntimeTask(String taskId, String leaseToken) throws Exception {
        return post("/api/v1/runtime/tasks/heartbeat", "{\"task_id\":\"" + taskId + "\",\"lease_token\":\"" + leaseToken + "\"}");
    }
    public String releaseRuntimeTask(String taskId, String leaseToken, String errorMessage) throws Exception {
        return post("/api/v1/runtime/tasks/release", "{\"task_id\":\"" + taskId + "\",\"lease_token\":\"" + leaseToken + "\",\"error_message\":\"" + jsonEscape(errorMessage) + "\"}");
    }
    public String requeueRuntimeTask(String taskId, int retryDelayMs, String errorMessage) throws Exception {
        return post("/api/v1/runtime/tasks/requeue", "{\"task_id\":\"" + taskId + "\",\"retry_delay_ms\":" + retryDelayMs + ",\"error_message\":\"" + jsonEscape(errorMessage) + "\"}");
    }
    public String getRuntimeTaskStats() throws Exception {
        return get("/api/v1/runtime/tasks/stats");
    }
    public String actionCallback(String callbackId, String status, String resultJson) throws Exception {
        return actionCallback(callbackId, status, resultJson, null);
    }
    public String actionCallback(String callbackId, String status, String resultJson, String callbackRequestId) throws Exception {
        return post("/api/v1/actions/callback", "{\"callback_id\":\"" + callbackId + "\",\"status\":\"" + status + "\",\"result\":" + resultJson + "}", callbackRequestId);
    }

    private String jsonEscape(String value) {
        if (value == null) {
            return "";
        }
        return value
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\r", "\\r")
            .replace("\n", "\\n");
    }
}
